Always exit on unrecoverable VM errors

oleschoenburg opened the pull request:
## Description

 * Sets a default uncaught exception handler that shuts down on any `VirtualMachineError`.
 * Shuts down on `VirtualMachineError`s in atomix threads.
 * Shuts down on `VirtualMachineError`s in actor threads.
 
<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

relates to #7807 but does not close it. 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

oleschoenburg commented:
The CI fails due to a timeout:
```
[2021-12-07T17:43:44.379Z] [INFO] Installing /home/jenkins/agent/workspace/ebe_7807-always-exit-on-vm-error/dist/target/camunda-cloud-zeebe-1.3.0-SNAPSHOT.tar.gz to /home/jenkins/agent/workspace/ebe_7807-always-exit-on-vm-error/m2-repository/io/camunda/camunda-cloud-zeebe/1.3.0-SNAPSHOT/camunda-cloud-zeebe-1.3.0-SNAPSHOT.tar.gz
[2021-12-07T17:48:37.563Z] Cancelling nested steps due to timeout
[2021-12-07T17:48:37.567Z] Sending interrupt signal to process
[2021-12-07T17:48:37.567Z] Killing processes
[2021-12-07T17:48:37.950Z] kill finished with exit code 0
[2021-12-07T17:48:51.002Z] Terminated
``` 

which is a bit suspicious but might just be i temporary issue in our CI environment. `mvn install` work on my machine ™

oleschoenburg commented:
I've incorporated your feedback into a new approach:

`FatalErrorHandler` should be easy to understand:

```
public final class FatalErrorHandler implements UncaughtExceptionHandler {
  private final Logger log;

  public FatalErrorHandler(final Logger log) {
    this.log = log;
  }

  public void handleError(final Throwable e) {
    if (e instanceof VirtualMachineError) {
      log.error("An unrecoverable VM error was thrown, exiting now.", e);
      System.exit(1);
    }
  }

  @Override
  public void uncaughtException(final Thread t, final Throwable e) {
    handleError(e);
  }
}
```
There is no rethrow functionality because it's not needed. Netty and Atomix manually call the `handleError` method to ensure that we properly handle VM errors. All other (normal) threads should do this implicitly because we use `FatalErrorHandler` as a default uncaught exception handler.

I still need to verify that this works exactly as I described by either manually testing or ideally writing tests for this.

npepinpe commented on util/src/main/java/io/camunda/zeebe/util/FatalErrorHandler.java:
```diff
@@ -0,0 +1,31 @@
+import java.lang.Thread.UncaughtExceptionHandler;
+import org.slf4j.Logger;
+
+public final class FatalErrorHandler implements UncaughtExceptionHandler {
```
:x: Let's document the behavior as any thing that just "exits" is probably counter intuitive.

npepinpe replied:
Let's also document what is currently considered a fatal error and why, I can see that getting lost over time :smile:

npepinpe commented on util/src/main/java/io/camunda/zeebe/util/FatalErrorHandler.java:
```diff
@@ -0,0 +1,31 @@
+    this.log = log;
+  }
+
+  public void handleError(final Throwable e) {
```
:x: I normally wouldn't push for changing method names, but in this case since we do exit (which is quite extreme), I would keep it in the name so it's very clear what happens when you use it :smile: wdyt?

oleschoenburg replied:
I agree that we should document this but I slightly disagree that we have to reflect that in the method name. As a user of `FatalErrorHandler`, I'd like to express: "Okay, some error happened, please make a decision if anything should be done about this" That's just my perspective and the reason why I kept the method name so generic. I can understand your point though, so your decision.

npepinpe replied:
That's fair, I just think exiting is a pretty extreme side effect and typically unexpected, so I would save the surprise by simply naming it `exitOnFatalError` or `exitIfFatalError` or some such, so a reader wouldn't even have to look into the implementation (except maybe to know what is or isn't a fatal error). But I'm happy to be proven wrong and maybe it's not needed, we can go with this and modify it if it does cause confusion :slightly_smiling_face:

npepinpe reviewed (changes requested):
:+1: Looks good. I have one request (split in two but it's basically the same), and one question: how will we test this? Can we even test this? I admit I didn't spend too much time thinking about it, but I'm curious what you came up with. Happy to do that with you too, ofc.

npepinpe reviewed (approved):
:+1: 

Before merging, please document the `FatalErrorHandler`, primarily why we consider `VirtualMachineError` unrecoverable and why we decide to exit because of it. No need for another review (but feel free to request one if you'd like)

npepinpe commented on util/src/main/java/io/camunda/zeebe/util/FatalErrorHandler.java:
```diff
@@ -0,0 +1,31 @@
+  public void handleError(final Throwable e) {
+    if (e instanceof VirtualMachineError) {
+      log.error("An unrecoverable VM error was thrown, exiting now.", e);
+      System.exit(1);
```
:wrench: Maybe we want to use a specific error code to indicate OOM? 137 is the one used by the Linux OOM killer (or 128 + 9). We could reuse this, or we could add one. My thoughts here are that when a user specifies "hey the program exited with X", then we know for sure it's because of this. I know we have the log statement of course, but having a specific status code can be useful if you want to react automatically to it. Let me know what you think.

oleschoenburg replied:
I like to idea to use a custom error code. I'd prefer not to use `137` for OOM precisely because it is so commonly used. I believe that this would make it a bit more difficult to distinguish between external OOM handling (the linux OOM killer, kubernetes etc.) and internal, JVM OOM handling. I'd suggest using exit code 156, which is ascii Z + B and as far as I can see only used by one product that isn't easily confused with Zeebe.

npepinpe commented on util/src/main/java/io/camunda/zeebe/util/FatalErrorHandler.java:
```diff
@@ -0,0 +1,31 @@
+
+  public void handleError(final Throwable e) {
+    if (e instanceof VirtualMachineError) {
+      log.error("An unrecoverable VM error was thrown, exiting now.", e);
```
:wrench: Operators are unlikely to read our Javadoc, so we might want to expand the error message to explain why this class of error is unrecoverable and why we decide to exit. If we had a public troubleshooting guide we'd probably want to link it here as well, but I guess there's no user facing doc to link here :disappointed: I'm open to any ideas or suggestions you have here though. But I'm just thinking, if I'm an SRE and I'm deploying a Zeebe cluster, and I run into this, do I know what I have to do?

npepinpe commented:
@oleschoenburg - check in with Roman about this. He had some good ideas about better error handling for actors and we might want to update this PR now or after with those.

npepinpe commented:
When I talked to Roman, we also discussed notifying actors on non fatal throwables. I think this is done in ActorJob, not ActorThread. Did you two discuss this, and if so, is there a reason this was discarded? Would we do this as a follow up? Doing this (or maybe instead closing the actor) would be one step towards a supervision like error handling strategy.

oleschoenburg commented:
I don't think we have talked about this, no. If I understand correctly, the current situation is that actors get notified about all `Exception`s (but not `Throwable`s) as long as they are not currently starting or closing.

npepinpe commented:
Not a blocker, we can still proceed here and add it later.

romansmirnov commented:
@oleschoenburg, sorry, seems that I was not too explicit about it, but actually, we touched on handling non-fatal throwables in the actor. As a concrete example, we talked about closing the log stream, and then we tried to understand what will happen in a failure case during the starting phase and during the runtime phase (we tried to understand what the failure listeners are doing, and then figured out the Zeebe partition will trigger a step down in Raft but also try to recover and Zeebe-side). However, on any failure (`Exceptions` and non-fatal `Throwables`), the log stream must be closed during the starting phase, and in addition, the appender future needs to be completed exceptionally. Otherwise, the log stream would be in an unclear state and the stream processor would still be waiting for the completion of the future. Basically, if something went wrong the actor should have a chance to handle it gracefully so that the entire partition can transition in an expected state and can try to recover accordingly.

In addition, to what you already have implemented, I would suggest that the `ActorJob` catches unhandled non-fatal `Throwables` in the same way as the `ActorJob` already does with regular uncaught `Exceptions`:

https://github.com/camunda-cloud/zeebe/blob/df9710cc95436a9b512ca751114a1e5fbdc7c685/util/src/main/java/io/camunda/zeebe/util/sched/ActorJob.java#L46-L48

That way, the log stream can handle those scenarios gracefully and close as mentioned above. And the `ActorThread` does not die.

Let's synchronize on this tomorrow to ensure that we are on the same page :)

oleschoenburg commented:
> sorry, seems that I was not too explicit about it, but actually, we touched on handling non-fatal throwables in the actor

No that was my bad, I wasn't connecting the dots here and I left out that part of our discussion because I wanted to understand the Logstream error handling first. I agree that extending the error handling in ActorJob would make sense.

romansmirnov reviewed (commented):
@oleschoenburg, I looked especially at the handling of non-fatal throwables in the actor. To me, it looks good. So, implementation-wise there is nothing to add. The only concern is about potential test cases though. I could imagine that it should be possible to write test cases for the actor to ensure that throwables are handled accordingly. WDYT?

npepinpe commented:
FYI, there is a way to test methods which also call `System.exit` by overriding the default `SecurityManager` and changing the default behavior to throw an exception instead of allowing the system to exit.

Unfortunately, `SecurityManager` is deprecated and marked for removal in 18, with no replacement. There's an open issue to introduce an API for this as several IDEs rely on this, but I have no idea if it's being used or not. Here's what a simple unit test would look like:

```java
package io.camunda.zeebe.util;

import static org.assertj.core.api.Assertions.assertThatCode;

import java.security.Permission;
import java.util.stream.Stream;
import java.util.zip.ZipError;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.slf4j.Logger;

@ExtendWith(MockitoExtension.class)
final class FatalErrorHandlerTest {
  private final ExitTrappingSecurityManager exitTrappingSecurityManager =
      new ExitTrappingSecurityManager();
  private SecurityManager previousSecurityManager;
  private @Mock Logger logger;

  @BeforeEach
  void beforeEach() {
    previousSecurityManager = System.getSecurityManager();
    System.setSecurityManager(exitTrappingSecurityManager);
  }

  @AfterEach
  void afterEach() {
    System.setSecurityManager(previousSecurityManager);
  }

  @ParameterizedTest
  @MethodSource("provideFatalErrors")
  void shouldExitOnFatalError(final Throwable fatalError) {
    // given
    final FatalErrorHandler handler = new FatalErrorHandler(logger);

    // when - then
    assertThatCode(() -> handler.handleError(fatalError))
        .isInstanceOf(ExitException.class)
        .hasFieldOrPropertyWithValue("status", FatalErrorHandler.EXIT_CODE);
  }

  private static Stream<Throwable> provideFatalErrors() {
    return Stream.of(
        new OutOfMemoryError(),
        new InternalError(),
        new StackOverflowError(),
        new UnknownError(),
        new ZipError("Failure"));
  }

  private static final class ExitException extends SecurityException {
    private final int status;

    public ExitException(final int status) {
      super();
      this.status = status;
    }

    public int getStatus() {
      return status;
    }
  }

  private static final class ExitTrappingSecurityManager extends SecurityManager {
    // it's necessary to override checkPermission in general, as otherwise you this would explode before we get to throw our exception in checkExit
    @Override
    public void checkPermission(final Permission perm) {}

    @Override
    public void checkPermission(final Permission perm, final Object context) {}

    @Override
    public void checkExit(final int status) {
      throw new ExitException(status);
    }
  }
}
```

npepinpe commented:
One small thing - it seems `InternalError` is covered here, correct? We know that `InternalError` is the error that is thrown when operations fail with `MappedByteBuffer` - for example, if the underlying file cannot be grown because there's no disk space anymore. See https://github.com/camunda-cloud/zeebe/issues/7607.

So with this, does it mean that the application would exit when this happens?

oleschoenburg commented:
@npepinpe Yes, that's correct. `InternalError` is a `VirtualMachineError` so we exit.

npepinpe commented:
Then I think we need to explicitly handle it in the Journal, at least until https://github.com/camunda-cloud/zeebe/issues/7607 is done. wdyt?

oleschoenburg commented:
@npepinpe How would we recover from this in the journal? If we are out of disk there is not much we can do, right?

npepinpe commented:
Hm, I suppose we can't compact without a new snapshot =/ I guess the user would have to manually expand the disk or make space? In Camunda Cloud, our disk will auto resize on the fly (up to a certain limit), so it would somewhat recover.

oleschoenburg commented:
That sounds to me like the behavior in this PR is ok then. If some hypervisor decides to restart the broker after exiting, that leaves a good opportunity to reattach resized disks (if I recall correctly not all managed k8s distributions can resize mounted disks, some need to detach the disk first).

oleschoenburg commented:
As discussed with @npepinpe I've removed all netty-related changes from this PR. As such, it doesn't actually close #7807 but it's still an improvement because VM errors in actors and atomix threads are handled correctly.

I've also rebased on develop and squashed the remaining commits as they were not providing any useful history.

npepinpe reviewed (approved):
:+1: 

Will you create a follow up, or update the other issue on what's left? I was also under the impression we'd discussed further tests for the actor scheduler, but I also can't remember anymore :smile:

oleschoenburg commented:
> I was also under the impression we'd discussed further tests for the actor scheduler, but I also can't remember anymore

Let's see what we could test:
* `ActorJob#invoke` throws ->`ActorJob`'s `FatalErrorHandler` is notified
* `ActorTask#execute` throws -> `ActorThread`'s `FatalErrorHandler` is notified
* `ActorThread` throws ->  default `UncaughtExceptionHandler` is notified.

I think this is also what we discussed before. As this would only test that we don't remove code and it would require additional constructors to inject spy-able `FatalErrorHandler`s I'm a bit hesitant to add it. @npepinpe I think you were in favor of this anyway, right?

npepinpe commented:
I was in favor because removing it would go entirely unnoticed, and would invalidate assumptions in the future about the system behavior, but perhaps the name is ominous enough that nobody would randomly remove it :smile: Let's go with the current state :+1:

oleschoenburg commented:
bors r+

oleschoenburg commented:
bors retry

oleschoenburg commented:
/backport
